web/static/js/components/import_export_tray.js

```diff
@@ -0,0 +1,161 @@
+  constructor (props) {
+    super(props)
+
+    this.state = {
+      pageContents: this.props.pageContents,
+      pageSettings: this.props.pageSettings,
+      stringifiedExportData: '',
+      stringifiedImportData: '',
+      exportDataButtonText: 'Copy to Clipboard'
+    }
+
+    this.onExportCopy = this.onExportCopy.bind(this)
+    this.stringifyExportData = this.stringifyExportData.bind(this)
+    this.onImportDataPaste = this.onImportDataPaste.bind(this)
+    this.onImportApply = this.onImportApply.bind(this)
+  }
+
+  componentWillReceiveProps (nextProps) {
+    this.setState({
+      pageContents: nextProps.pageContents,
+      pageSettings: nextProps.pageSettings,
+      stringifiedExportData: '',
+      stringifiedImportData: '',
+      exportDataButtonText: 'Copy to Clipboard'
+    })
+  }
+
+  componentDidMount () {
+    this.stringifyExportData()
+  }
+
+  componentDidUpdate (prevProps, prevState) {
+    if (this.state.stringifiedExportData !== '') return
+
+    this.stringifyExportData()
+  }
+
+  onExportCopy () {
+    this.setState({exportDataButtonText: 'Copied!'})
+
+    clearTimeout(window.exportDataButtonTextTimeout)
+    window.exportDataButtonTextTimeout = setTimeout(() => {
+      this.setState({exportDataButtonText: 'Copy to Clipboard'})
+    }, 2000)
+  }
+
+  stringifyExportData () {
+    let exportData = {
+      pageSettings: Object.assign({}, this.state.pageSettings, {origin: window.location.origin}) ,
+      pageContents: this.state.pageContents
+    }
+
+    const stringifiedExportData = JSON.stringify(exportData)
+
+    this.setState({
+      stringifiedExportData: stringifiedExportData
+    })
+  }
+
+  onImportDataPaste (e) {
+    this.setState({stringifiedImportData: e.target.value})
+  }
+
+  onImportApply () {
+    this.props.updateImportProgress({type: 'in-progress', text: 'Importing content. This may take some time.'})
+
+    const data = this.getImportData()
+
+    if (data) {
+      this.props.importData(data.pageSettings, data.pageContents)
+    } else {
+      this.setImportProgressError({type: 'error', text: 'The data entered is invalid.'})
+    }
+  }
+
+  getImportData () {
+    const importData = this.parseStringifiedImportData()
+    return (importData && importData.pageSettings && importData.pageContents) ? importData : null
+  }
+
+  parseStringifiedImportData () {
+    try {
+      return JSON.parse(this.state.stringifiedImportData)
+    } catch (ex) {
+      return null
+    }
+  }
+
+  setImportProgressError (error) {
+    this.props.updateImportProgress(error)
+
+    clearTimeout(window.importProgressErrorTimeout)
+    window.importProgressErrorTimeout = setTimeout(() => {
+      this.props.updateImportProgress(null)
+      this.setState({stringifiedImportData: ''})
+    }, 3500)
+  }
+
+  render () {
+    if (this.props.importProgress) {
+      return this.renderImportStatus()
+    } else {
+      return this.renderImportExportActions()
+    }
+  }
+
```

Comment:
this render features a hidden textarea that we prefill with data and use the CopyToClipboard wrapper